### ISSUE TITLE
Tests for new bam datatypes 

### ISSUE THREAD

**Issue by jennaj** [galaxyproject/usegalaxy-playbook#102]
**18.01** Test changes re: new bam datatypes https://docs.galaxyproject.org/en/master/releases/18.01_announce.html

**Test history** https://usegalaxy.org/u/jen/h/test-history-bam-datatypes

**Upload done:**
- [x] **bam** -- only assigns to coord-sorted bams with intact EOF
- [x] **unsorted.bam** https://github.com/galaxyproject/galaxy/pull/5674 -- currently auto-detect assigns to any bam with no EOF, but should really fail Upload instead. The distinct ticket is here https://github.com/galaxyproject/usegalaxy-playbook/issues/103.

**Upload hold:** -- no tools produce these datatypes in Galaxy intact. Will retest upload/upgraded tools at the same time for these two.
- [x] **qname_sorted.bam**
- [ ] **qname_input_sorted.bam**

**Tools:**
_add tools/functions here once upgraded_

With respect to ^^ 

- **Upload** handles complete SAM/BAM uploads correctly but not incomplete BAM uploads for some cases. Details in comments below with ticket links.
 - **SAM-to-BAM** will produce a green success `bam` dataset even without a proper RG/SO header. Example in https://github.com/galaxyproject/tools-iuc/issues/1774
 -  **Samtools sort** should output `qname_sorted.bam`. Currently creates failed-green unindexed BAMs (a bug, no ticket yet). Tool help text update included in this. Re-test then ticket.
 - **Picard sort** should output `qname_sorted.bam`. Currently, a queryname sort outputs SAM, could have an option for BAM. Tool help text update included in this. Re-test then ticket.
 - **Support FAQs need an update** 

Moved from https://github.com/galaxyproject/usegalaxy-playbook/issues/86

**Comment by jennaj** [comment]
Issue opened for tool upgrades: https://github.com/galaxyproject/tools-iuc/issues/1774

**Comment by jennaj** [comment]
Using the original sorted BAMs in the same test history, Upload correctly assigns BAM/SAM datatypes. And maybe we should consider adding in a `qname_sorted.sam` datatype??

<img width="630" alt="screen shot 2018-03-15 at 7 59 25 pm" src="https://user-images.githubusercontent.com/8999735/37501446-b0f69962-288b-11e8-99f6-69a71e5e18c7.png">
<img width="556" alt="screen shot 2018-03-15 at 7 59 38 pm" src="https://user-images.githubusercontent.com/8999735/37501447-b14d95be-288b-11e8-88ee-48529b435319.png">

**Comment by jennaj** [comment]
All looks good now as far as Upload goes. Great!

Labeled results. Partially uploaded BAMs and Metadata-problem Qname sorted BAMs are the odd cases -- but not an emergency.

<img width="248" alt="screen shot 2018-03-16 at 8 02 30 pm" src="https://user-images.githubusercontent.com/8999735/37551071-f8e938ae-2955-11e8-9dc7-8f150d172332.png">

<img width="251" alt="screen shot 2018-03-16 at 8 07 42 pm" src="https://user-images.githubusercontent.com/8999735/37551072-f9038aba-2955-11e8-83af-f2b0d7fe5d6e.png">

<img width="246" alt="screen shot 2018-03-16 at 8 08 27 pm" src="https://user-images.githubusercontent.com/8999735/37551073-f91d8e2e-2955-11e8-8460-e1b6cd191d1c.png">

**Comment by jennaj** [comment]
I retested the bam/sam new datatypes with the Upload tool on https://usegalaxy.org with the 18.05 pre-releaes installed.

ping @dannon @guerler 

Related discussion for reference: https://github.com/galaxyproject/galaxy/issues/5889

Test history same as above https://usegalaxy.org/u/jen/h/bam-upload, new tests start with dataset 53 and made use of sequence-only extracted BAMs from NCBI tools, plus an intact/correct coordinate sorted BAM from HISAT2 (dataset 4).

Few unexpected results occurred. Tested datatype assignments for 1) complete uploads 2) truncated BAMs/SAMs (truncated == not fully loaded via FTP). Both for sequence-only and mapped alignment-included data.

**Test 1**

Dataset 53 is the BAM extracted from NCBI (sequences only, no header).
Dataset 54 is the SAM extracted from NCBI (sequences only, no header).

**Unexpected results for Test 1**

Both datasets are "green/success" with `bam/sam` datatype assigned correctly, and the download looks complete (even though the output from the NCBI tool has sequence only no header) but if I click on the eye icon, this message is reported instead of a preview:

```
QNAME | FLAG | RNAME | POS | MAPQ | CIGAR | MRNM | MPOS | ISIZE | SEQ | QUAL | OPT
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Could not display BAM file, error was:
file has no sequences defined (mode='rb') - is it SAM/BAM format? Consider opening with check_sq=False
```

Ran SortSAM on 53/54, both qname and coord sort options, and both technically "successful" but don't produce usable `sam` or `bam` datasets. Tests in datasets 55-59. *Shouldn't the tool report something like **"no alignments found in BAM, data could not be sorted" and fail instead**?* 

**Test 2**

Downloaded and uploaded (autodetect datatype) datasets 53 (`bam`), only sequence no header, and 54 (`bam`) only sequence no header. Both full uploads and truncated uploads. Tests in datsets 60-65.

**Unexpected results for Test 2**

Dataset 60, a complete upload of dataset 53 (bam), only sequence no header, also appears intact in UI, datatype detected as `bam`, but clicking on th eye gives this message:

```
Could not display BAM file, error was:
--
file has no sequences defined (mode='rb') - is it SAM/BAM format? Consider opening with check_sq=False

QNAME | FLAG | RNAME | POS | MAPQ | CIGAR | MRNM | MPOS | ISIZE | SEQ | QUAL | OPT
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Could not display BAM file, error was:
file has no sequences defined (mode='rb') - is it SAM/BAM format? Consider opening with check_sq=False

```
Dataset 61, an truncated upload of dataset 53 (`bam`), sequence only no header, also appears intact in UI (primary dataset), datatype detected as `bam_native`, but clicking on th eye gives this message:

```
QNAME | FLAG | RNAME | POS | MAPQ | CIGAR | MRNM | MPOS | ISIZE | SEQ | QUAL | OPT
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Could not display BAM file, error was:
no BGZF EOF marker; file may be truncated
```

_**Note**: a secondary hidden dataset 61 (red) is not revealed in the history yet, downstream tools need to be run to reveal that. See Test 3._ 

**Test 3**

Dataset 62, a SortSam on dataset 60 (`bam`), sequences only, no header, also appears intact in UI, datatype detected as `bam`. 

**Unexpected results for Test 3**

Clicking on the eye for dataset 62 gives this message:

```
QNAME | FLAG | RNAME | POS | MAPQ | CIGAR | MRNM | MPOS | ISIZE | SEQ | QUAL | OPT
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Could not display BAM file, error was:
file has no sequences defined (mode='rb') - is it SAM/BAM format? Consider opening with check_sq=False
```

_Similiar to **Issue 1**._ No header/alignments, so cannot really be sorted. Should fail, yes?

Dataset 63, a SortSam on dataset 61 (`sam`), sequences only no header, remains as a paused job and the hidden error (red) 61 dataset is now revealed to the user in the history UI.

_Confusing for users..._

**Test 4**

Datasets 64 and 65 (`sam`), no header sequences only, Uploaded both complete and truncated. Used autodetect for datatype. A truncated `sam` would only cause problems with tools if truncated in the middle of a record, but still seemed worth testing both.

**Unexpected results for Test 4**

Both are given the datatype `otu_map` as detected by the Upload tool (wierd!). Clicking on the eye will display the data, but the column headers are _not_ for `sam` format, are for `otu_map`.

Tools that expect `sam` or `bam` input, like those that can be used to convert Bam/Sam-to-Fastq, do not recognize these data as inputs due to incorrect datatype assignment. Example: Bamtools >> Convert from BAM to FastQ (Galaxy Version 2.27.0.0). Confusing for users - they could to change the datatype to `sam`, but then later will run into problems like the ones above, since are not header-intact `sam` data.

_Maybe we need a **distinct `bam` and `sam` datatype when those files just contain sequences no header**? And have the Upload tool recognize and assign those._

**Test 5** 

Datasets 66 and 67, a SortSAM on the intact coordinate sorted dataset 4 (`bam`), using both queryname and coordinate sort work. Expected - just checking to make sure functional!

**Unexpected results for Test 5** 

None! Good news!

**Test 6**

Dataset 68, 71, 72: a download and complete reupload of Dataset 4 (coordinate sorted `bam` from HISAT2), then ran SortSam with queryname and coord sort options.

**Unexpected results for Test 6**

None! Also, good news! Is clear that things work when data is intact with expected content. It is when the data is "off" that odd problems happen, and it is difficult for users to know if data is correct or "off" until downstream tools are run against them. Not catching problems format problems genenates more bug reports and user frustration...

**Test 7**

Dataset 73, downloaded and uploaded Dataset 66 (`bam`), coordinate sorted Dataset 4 with SortSam, with alignments and a complete header.

**Unexpected results for Test 7**

None, worked as expected!

**Test 8**

Dataset 74, downloaded and uploaded Dataset 67 (`sam`), queryname sorted Dataset 4 with SortSam, with alignments and a complete header.

**Unexpected results for Test 8**

None I think. Autodetected datatype `sam`. A qname sort sam datatype would be more informative. This was already brought up, but maybe not added yet as of 18.05.

**Test 9**

Dataset 75, downloaded and uploaded Dataset 67 queryname sorted Dataset 4 with SortSam, with alignments and a complete header. I downloaded the `sam`, then converted to `bam` with samtools view. The original "SO:queryname" is in the bam.

**Unexpected results for Test 9**

With Upload, autodetect assigned the datatype `unsorted.bam`. Shouldn't this now be `qname_sorted.bam` as of 18.01 through to 18.05? Or are those datatypes/autodetect methods still a work in progress (placeholders for later)?

Not sure which or any we want to address, or think we have addressed but there are problems, in 18.05.

**Comment by jennaj** [comment]
Passes now, see ticket: https://github.com/galaxyproject/usegalaxy-playbook/issues/174

Note: `qname_input_sorted.bam` not tested. Not sure if output by any tools yet, but it doesn't get assigned during upload (correctly assigning datatype in upload was the primary concern)

Closing
